File: ios/OpuslibModule.swift
import ExpoModulesCore
import AVFoundation

/**
 * OpuslibModule - Expo module for Opus 1.6 audio encoding with DRED support
 *
 * This module provides native audio capture and Opus 1.6 encoding with Deep Redundancy (DRED)
 * for improved audio quality on lossy networks.
 */
public class OpuslibModule: Module {
  private var audioEngineManager: AudioEngineManager?
  private var isStreaming = false

  public func definition() -> ModuleDefinition {
    Name("Opuslib")

    // Events
    Events("audioChunk", "amplitude", "audioStarted", "audioEnd", "error")

    // Start streaming method
    AsyncFunction("startStreaming") { (config: AudioConfig) in
      try self.startStreaming(config: config)
    }

    // Stop streaming method
    AsyncFunction("stopStreaming") {
      try await self.stopStreaming()
    }

    // Pause streaming method
    Function("pauseStreaming") {
      self.pauseStreaming()
    }

    // Resume streaming method
    Function("resumeStreaming") {
      self.resumeStreaming()
    }
  }

  // MARK: - Public Methods

  private func startStreaming(config: AudioConfig) throws {
    print("[OpuslibModule] 🎬 startStreaming() called with DRED: \(config.dredDuration ?? 100)ms")

    guard !isStreaming else {
      throw AudioStreamError.alreadyStreaming
    }

    // Request microphone permission
    print("[OpuslibModule] 🔐 Requesting microphone permission...")
    try requestMicrophonePermission()
    print("[OpuslibModule] ✅ Microphone permission granted")

    // Create audio engine manager
    print("[OpuslibModule] 🏗️ Creating AudioEngineManager...")
    let manager = AudioEngineManager(config: config)
    print("[OpuslibModule] ✅ AudioEngineManager created")

    // Set up event callbacks — audioStarted/audioEnd come from encoding thread
    manager.setOnAudioChunk { [weak self] data, timestamp, sequenceNumber, audioLevel in
      self?.sendEvent("audioChunk", [
        "data": data,
        "timestamp": timestamp,
        "sequenceNumber": sequenceNumber,
        "audioLevel": audioLevel
      ])
    }

    manager.setOnStarted { [weak self] timestamp, sampleRate, channels, bitrate, frameSize, preSkip in
      self?.sendEvent("audioStarted", [
        "timestamp": timestamp,
        "sampleRate": sampleRate,
        "channels": channels,
        "bitrate": bitrate,
        "frameSize": frameSize,
        "preSkip": preSkip
      ])
    }

    manager.setOnEnd { [weak self] timestamp, totalDuration, totalPackets in
      self?.sendEvent("audioEnd", [
        "timestamp": timestamp,
        "totalDuration": totalDuration,
        "totalPackets": totalPackets
      ])
    }

    manager.setOnAmplitude { [weak self] rms, peak, timestamp in
      self?.sendEvent("amplitude", [
        "rms": rms,
        "peak": peak,
        "timestamp": timestamp
      ])
    }

    manager.setOnError { [weak self] error in
      self?.sendEvent("error", [
        "code": "AUDIO_ENGINE_ERROR",
        "message": error.localizedDescription
      ])
    }

    // Start audio capture + encoding
    try manager.start()

    audioEngineManager = manager
    isStreaming = true

    print("[OpuslibModule] ✅ Started streaming")
  }

  private func stopStreaming() async throws {
    guard isStreaming else {
      return
    }

    // stop() triggers flushAndStop() which emits audioEnd from encoding thread
    audioEngineManager?.stop()
    audioEngineManager = nil
    isStreaming = false

    print("[OpuslibModule] Stopped streaming")
  }

  private func pauseStreaming() {
    guard isStreaming else {
      return
    }

    audioEngineManager?.pause()
    print("[OpuslibModule] Paused streaming")
  }

  private func resumeStreaming() {
    guard isStreaming else {
      return
    }

    audioEngineManager?.resume()
    print("[OpuslibModule] Resumed streaming")
  }

  // MARK: - Private Methods

  private func requestMicrophonePermission() throws {
    let audioSession = AVAudioSession.sharedInstance()

    switch audioSession.recordPermission {
    case .granted:
      return

    case .denied:
      throw AudioStreamError.permissionDenied

    case .undetermined:
      var permissionGranted = false
      let semaphore = DispatchSemaphore(value: 0)

      audioSession.requestRecordPermission { granted in
        permissionGranted = granted
        semaphore.signal()
      }

      semaphore.wait()

      if !permissionGranted {
        throw AudioStreamError.permissionDenied
      }

    @unknown default:
      throw AudioStreamError.permissionDenied
    }
  }
}

// MARK: - Configuration

/**
 * Audio configuration for Opus encoding
 */
struct AudioConfig: Record {
  @Field var sampleRate: Int = 16000
  @Field var channels: Int = 1
  @Field var bitrate: Int = 24000
  @Field var frameSize: Double = 20.0
  @Field var packetDuration: Double = 20.0
  @Field var dredDuration: Int? = 100  // NEW: DRED recovery duration in ms
  @Field var enableAmplitudeEvents: Bool? = false
  @Field var amplitudeEventInterval: Double? = 16.0
  @Field var audioLevelWindow: Int? = 360  // RMS window duration in ms (default 360)
  @Field var saveDebugAudio: Bool? = false
}

// MARK: - Errors

/**
 * Custom errors for audio streaming
 */
enum AudioStreamError: Error {
  case alreadyStreaming
  case notStreaming
  case permissionDenied
  case audioEngineError(String)
  case opusEncodingError(String)
}

File: android/src/main/java/expo/modules/opuslib/OpuslibModule.kt
package expo.modules.opuslib

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

/**
 * OpuslibModule - Expo module for Opus 1.6 audio encoding with DRED support
 *
 * This module provides native audio capture and Opus 1.6 encoding with Deep Redundancy (DRED)
 * for improved audio quality on lossy networks.
 */
class OpuslibModule : Module() {
  private var audioRecordManager: AudioRecordManager? = null
  private var isStreaming = false

  override fun definition() = ModuleDefinition {
    Name("Opuslib")

    // Events
    Events("audioChunk", "amplitude", "audioStarted", "audioEnd", "error")

    // Start streaming method
    AsyncFunction("startStreaming") { config: AudioConfig ->
      startStreaming(config)
    }

    // Stop streaming method
    AsyncFunction("stopStreaming") {
      stopStreaming()
    }

    // Pause streaming method
    Function("pauseStreaming") {
      pauseStreaming()
    }

    // Resume streaming method
    Function("resumeStreaming") {
      resumeStreaming()
    }
  }

  // MARK: - Public Methods

  private fun startStreaming(config: AudioConfig) {
    android.util.Log.d(TAG, "🎬 startStreaming() called with DRED: ${config.dredDuration}ms")

    if (isStreaming) {
      throw AudioStreamException("ALREADY_STREAMING", "Audio streaming is already active")
    }

    // Check microphone permission
    val context = appContext.reactContext ?: throw AudioStreamException(
      "NO_CONTEXT",
      "React context not available"
    )

    if (ContextCompat.checkSelfPermission(
        context,
        Manifest.permission.RECORD_AUDIO
      ) != PackageManager.PERMISSION_GRANTED
    ) {
      throw AudioStreamException(
        "PERMISSION_DENIED",
        "Microphone permission not granted"
      )
    }

    android.util.Log.d(TAG, "✅ Microphone permission granted")

    // Create audio record manager
    android.util.Log.d(TAG, "🏗️ Creating AudioRecordManager...")
    val manager = AudioRecordManager(context, config)
    android.util.Log.d(TAG, "✅ AudioRecordManager created")

    // Set up event callbacks — audioStarted/audioEnd come from encoding thread
    android.util.Log.d(TAG, "🔗 Setting up event callbacks...")
    manager.setOnAudioChunk { data, timestamp, sequenceNumber, audioLevel, duration, frameCount ->
      sendEvent("audioChunk", mapOf(
        "data" to data,
        "timestamp" to timestamp,
        "sequenceNumber" to sequenceNumber,
        "audioLevel" to audioLevel,
        "duration" to duration,
        "frameCount" to frameCount
      ))
    }

    manager.setOnStarted { timestamp, sampleRate, channels, bitrate, frameSize, preSkip ->
      sendEvent("audioStarted", mapOf(
        "timestamp" to timestamp,
        "sampleRate" to sampleRate,
        "channels" to channels,
        "bitrate" to bitrate,
        "frameSize" to frameSize,
        "preSkip" to preSkip
      ))
    }

    manager.setOnEnd { timestamp, totalDuration, totalPackets ->
      sendEvent("audioEnd", mapOf(
        "timestamp" to timestamp,
        "totalDuration" to totalDuration,
        "totalPackets" to totalPackets
      ))
    }

    manager.setOnAmplitude { rms, peak, timestamp ->
      sendEvent("amplitude", mapOf(
        "rms" to rms,
        "peak" to peak,
        "timestamp" to timestamp
      ))
    }

    manager.setOnError { error ->
      android.util.Log.e(TAG, "❌ Error from manager: ${error.message}")
      sendEvent("error", mapOf(
        "code" to "AUDIO_RECORD_ERROR",
        "message" to error.message
      ))
    }

    // Start audio capture + encoding
    android.util.Log.d(TAG, "🚀 Calling manager.start()...")
    manager.start()
    android.util.Log.d(TAG, "✅ manager.start() completed")

    audioRecordManager = manager
    isStreaming = true

    android.util.Log.d(TAG, "✅ Started streaming")
  }

  private fun stopStreaming() {
    if (!isStreaming) {
      return
    }

    // stop() triggers flushAndStop() which emits audioEnd from encoding thread
    audioRecordManager?.stop()
    audioRecordManager = null
    isStreaming = false

    android.util.Log.d(TAG, "Stopped streaming")
  }

  private fun pauseStreaming() {
    if (!isStreaming) {
      return
    }

    audioRecordManager?.pause()
    android.util.Log.d(TAG, "Paused streaming")
  }

  private fun resumeStreaming() {
    if (!isStreaming) {
      return
    }

    audioRecordManager?.resume()
    android.util.Log.d(TAG, "Resumed streaming")
  }

  companion object {
    private const val TAG = "OpuslibModule"
  }
}

// MARK: - Configuration

/**
 * Audio configuration for Opus encoding
 */
class AudioConfig : Record {
  @Field
  var sampleRate: Int = 16000

  @Field
  var channels: Int = 1

  @Field
  var bitrate: Int = 24000

  @Field
  var frameSize: Double = 20.0

  @Field
  var packetDuration: Double = 20.0

  @Field
  var dredDuration: Int = 100  // NEW: DRED recovery duration in ms

  @Field
  var enableAmplitudeEvents: Boolean = false

  @Field
  var amplitudeEventInterval: Double = 16.0

  @Field
  var audioLevelWindow: Int = 360  // RMS window duration in ms (default 360)

  @Field
  var saveDebugAudio: Boolean = false
}

// MARK: - Errors

/**
 * Custom exception for audio streaming errors
 */
class AudioStreamException(code: String, message: String) : CodedException(code, message, null)
